debian/ubuntu-advantage-tools.preinst

```diff
@@ -2,100 +2,143 @@
 
 set -e
 
-remove_old_config_fields() {
-    PREVIOUS_PKG_VER="$1"
-    if dpkg --compare-versions "$PREVIOUS_PKG_VER" le "27.8"; then
-        if grep -q "^license_check_log_file:" /etc/ubuntu-advantage/uaclient.conf; then
-            sed -i '/^license_check_log_file:.*$/d' /etc/ubuntu-advantage/uaclient.conf || true
-        fi
+
+migrate_user_config() {
+    if [ ! -f /etc/ubuntu-advantage/uaclient.conf ]; then
+        return
     fi
+
+    curr_conf_hash=$(md5sum /etc/ubuntu-advantage/uaclient.conf | awk '{print $1}')
+    expected_conf_hash=$(dpkg-query --showformat='${Conffiles}\n' --show ubuntu-advantage-tools | grep /etc/ubuntu-advantage/uaclient.conf | awk '{print $2}')
+
+    if [ "$curr_conf_hash" = "$expected_conf_hash" ]; then
+        # no change, just create a minimal user_config.json
+        echo "{}" > /var/lib/ubuntu-advantage/user-config.json
+    else
+        # Back up existing conffile in case of an error unwind
+        cp -a /etc/ubuntu-advantage/uaclient.conf /etc/ubuntu-advantage/uaclient.conf.preinst-remove
+        # This removes the "ua_config" sub-document from uaclient.conf and
+        # moves it to it's own json file in /var/lib/ubuntu-advantage/user-config.json.
+        # The fields under "ua_config" are intended to be set/unset by the
+        # user via the `pro config set` command, so we don't want them covered as a "conffile".
+        # It writes back a uaclient.conf that will match the new default
+        # uaclient.conf as long as no defaults have changed - avoiding
+        # an unnecessary conffile prompt.
+        # We print warning messages if errors occur with instructions to help the user resolve them.
+        /usr/bin/python3 -c "
+import json
```

Comment:
I think this fits better in a separate file, and then we just call it here?

Reply:
I would love that!
AFAICT it's not possible since we have to do this in preinst - but I would really like to be wrong here.

Reply:
@orndorffgrant This need to be in preinst to avoid the debconf prompt right ? If yes, what if on preinst we just create a backup of the conf file and bring back the default value. Then, on postinst we can perform the transition and move that logic into a separate module

Reply:
That's a great idea! Yes I'll do that